File: json/json-annotation-processor/src/test/java/ru/tinkoff/kora/json/annotation/processor/dto/DtoOnlyWriter.java
package ru.tinkoff.kora.json.annotation.processor.dto;

import com.fasterxml.jackson.core.JsonGenerator;
import ru.tinkoff.kora.json.common.annotation.JsonField;
import ru.tinkoff.kora.json.common.annotation.JsonSkip;
import ru.tinkoff.kora.json.common.annotation.JsonWriter;

import javax.annotation.Nullable;
import java.io.IOException;

@JsonWriter
public record DtoOnlyWriter(
    String field1,
    @JsonField("renamedField2") String field2,
    @JsonField(writer = FieldWriter.class) Inner field3,
    @JsonSkip String field4
) {

    public static final class FieldWriter implements ru.tinkoff.kora.json.common.JsonWriter<Inner> {
        @Override
        public void write(JsonGenerator gen, @Nullable Inner object) throws IOException {
            if (object == null) {
                gen.writeNull();
            } else {
                gen.writeString(object.value());
            }
        }
    }

    public record Inner(String value) {}
}

File: json/json-symbol-processor/src/test/kotlin/ru/tinkoff/kora/json/ksp/dto/DtoOnlyWriter.kt
package ru.tinkoff.kora.json.ksp.dto

import com.fasterxml.jackson.core.JsonGenerator
import ru.tinkoff.kora.json.common.annotation.JsonField
import ru.tinkoff.kora.json.common.annotation.JsonSkip
import ru.tinkoff.kora.json.common.annotation.JsonWriter

@JsonWriter
data class DtoOnlyWriter(
    val field1: String,
    @JsonField("renamedField2") val field2: String,
    @JsonField(writer = FieldWriter::class) val field3: Inner,
    @JsonSkip val field4: String
) {
    class FieldWriter : ru.tinkoff.kora.json.common.JsonWriter<Inner?> {
        override fun write(gen: JsonGenerator, obj: Inner?) {
            if (obj == null) {
                gen.writeNull()
            } else {
                gen.writeString(obj.value)
            }
        }
    }

    data class Inner(val value: String)
}
